cli/src/main/kotlin/tools/samt/cli/DiagnosticFormatter.kt

```diff
@@ -0,0 +1,506 @@
+        } else {
+            (severityBackgroundColor(severity))(text)
+        }
+
+        return if (withBold) {
+            bold(colorFormatted)
+        } else {
+            colorFormatted
+        }
+    }
+
+    private fun formatSeverityIndicator(severity: DiagnosticSeverity): String = when (severity) {
+        DiagnosticSeverity.Error -> formatTextForSeverity("ERROR:", severity, withBold = true)
+        DiagnosticSeverity.Warning -> formatTextForSeverity("WARNING:", severity, withBold = true)
+        DiagnosticSeverity.Info -> formatTextForSeverity("INFO:", severity, withBold = true)
+    }
+
+    private fun formatFilePathRelativeToWorkingDirectory(filePath: String): String {
+        val workingDirectory = diagnosticController.workingDirectoryAbsolutePath
+        return filePath.removePrefix(workingDirectory).removePrefix(File.separator)
+    }
+
+    private fun formatGlobalMessage(message: DiagnosticGlobalMessage): String = buildString {
+        append(formatSeverityIndicator(message.severity))
+        append(" ")
+        append(message.message)
+        appendLine()
+    }
+
+    private fun formatMessage(message: DiagnosticMessage, context: DiagnosticContext): String = buildString {
+
+        // <severity>: <message>
+        append(formatSeverityIndicator(message.severity))
+        append(" ")
+        append(message.message)
+        appendLine()
+
+        val errorSourceFilePath = if (message.highlights.isNotEmpty()) {
+            message.highlights.first().location.source.absolutePath
+        } else {
+            context.source.absolutePath
+        }
+
+        // -----> <file path>:<location>
+        append(gray(" ---> "))
+        append(formatFilePathRelativeToWorkingDirectory(errorSourceFilePath))
+        if (message.highlights.isNotEmpty()) {
+            val firstHighlight = message.highlights.first()
+            val firstHighlightLocation = firstHighlight.location
+            append(":")
+            append(firstHighlightLocation.toString())
+        }
+        appendLine()
+
+        if (message.highlights.isNotEmpty()) {
+            appendLine()
+            append(formatMessageHighlights(message))
+        }
+
+        // print highlight annotations
+        if (message.annotations.isNotEmpty()) {
+            appendLine()
+            appendLine(formatHighlightAnnotations(message.annotations))
+        }
+    }
+
+    private fun formatMessageHighlights(message: DiagnosticMessage): String = buildString {
+        val highlights = message.highlights
+        require(highlights.isNotEmpty())
+
+        // group highlights by source file
+        val mainSourceFileAbsolutePath = highlights.first().location.source.absolutePath
+        val highlightsBySourceFile = highlights.groupBy { it.location.source.absolutePath }
+
+        // print the highlights for the main source file
+        val mainSourceFileHighlights = highlightsBySourceFile[mainSourceFileAbsolutePath]!!
```

Comment:
Could be done without the `!!`
```suggestion
        val mainSourceFileHighlights = highlightsBySourceFile.getValue(mainSourceFileAbsolutePath)
```